Add CSS revert-rule keyword (#43808)
mdn/content: files/en-us/web/css/reference/values/revert-rule/index.md

---
title: revert-rule
slug: Web/CSS/Reference/Values/revert-rule
page-type: css-keyword
browser-compat: css.types.global_keywords.revert-rule
sidebar: cssref
---

> The **`revert-rule`** [CSS-wide keyword](/en-US/docs/Web/CSS/Reference/Values/Data_types#css-wide_keywords) rolls back the value of a property as if the current [style rule](/en-US/docs/Web/CSS/Syntax#css_rulesets) had not been present at all. The cascade then determines the value from the remaining declarations — this could be another rule in the same [cascade layer](/en-US/docs/Web/CSS/Reference/At-rules/@layer), a rule in a different layer, a different [style origin](/en-US/docs/Glossary/Style_origin), or a [defaulted value](/en-US/docs/Web/CSS/Guides/Cascade/Property_value_processing#defaulting) (inherited or initial).

Reviewer: ```suggestion
The **`revert-rule`** [CSS-wide keyword](/en-US/docs/Web/CSS/Reference/Values/Data_types#css-wide_keywords) rolls back the cascaded value of a property to the value it would have had if the current [style rule](/en-US/docs/Web/CSS/Syntax#css_rulesets) had not been present. The cascade then determines the value from the remaining declarations — this could be another rule in the same [cascade layer](/en-US/docs/Web/CSS/Reference/At-rules/@layer), a rule in a different layer, a different [style origin](/en-US/docs/Glossary/Style_origin), or a [default value](/en-US/docs/Web/CSS/Guides/Cascade/Property_value_processing#defaulting) (`inherited` or `initial`).
```

> The `revert-rule` keyword behaves like {{cssxref("revert-layer")}} in the animation origin.

Reviewer: do we have more context or link for "revert-layer in the animation origin"? Can we explain this a bit more?

This keyword can be applied to any CSS property, including the CSS shorthand property {{cssxref("all")}}.

## Revert-rule vs. revert-layer vs. revert

The `revert-rule`, {{cssxref("revert-layer")}}, and {{cssxref("revert")}} keywords each roll back the cascade, but at different levels of granularity:

- {{cssxref("revert")}} removes all declarations from the current [style origin](/en-US/docs/Glossary/Style_origin), rolling back to the previous origin (for example, from author styles to user-agent styles).
- {{cssxref("revert-layer")}} removes all declarations from the current [cascade layer](/en-US/docs/Web/CSS/Reference/At-rules/@layer), rolling back to the previous layer within the same origin.
- `revert-rule` removes only the declarations from the current style rule. Other rules in the same cascade layer still apply.

This makes `revert-rule` useful for conditionally ignoring specific declarations within a rule while still respecting declarations from other rules in the same layer.

## Examples

> ### Rolling back to the previous rule

Reviewer: ```suggestion
### Rolling back to the previous matching rule
```

> In this example, two rules target the same element. The second rule uses `revert-rule` on the `color` property, which causes the cascade to determine the value as if the `p.special` rule were not present, falling back to the value established by the first rule.

Reviewer: maybe we can add what would happen normally without `revert-rule`
I ended up rephrasing a bit more, see if you like it
```suggestion
In this example, two rules target the same element. The more specific `p.special` rule would normally win for the `color` property. However, the `revert-rule` keyword causes the cascade to determine the property value as if that rule were not present, falling back to the value established by the `p` rule.
```

#### HTML

```html
<p class="special">This paragraph has special styling.</p>
```

#### CSS

```css hidden
body {
  font-family: system-ui;
}

@supports not (color: revert-rule) {
  body::before {
    content: "Your browser doesn't support the revert-rule keyword yet.";
    background-color: wheat;
    display: block;
    text-align: center;
    padding: 1em;
  }
}
```

```css
p {
  color: blue;
  font-weight: bold;
}

p.special {
  color: revert-rule;
  border: 1px solid currentcolor;
}
```

#### Result

{{EmbedLiveSample('Rolling_back_to_the_previous_rule', '100%', 120)}}

> The paragraph text is blue (from the `p` rule), because the `color: revert-rule` declaration in the `p.special` rule causes it to be ignored for the `color` property. The `font-weight: bold` and `border` declarations are unaffected.

Reviewer: Moving around a few words to make the para compact
```suggestion
The paragraph text is blue from the `p` rule because `color: revert-rule` causes the `color` declaration in `p.special` to be ignored. The `font-weight` and `border` declarations are unaffected.
```

### Reverting from a style attribute

> When `revert-rule` is used in a [style attribute](/en-US/docs/Web/HTML/Reference/Global_attributes/style), it causes the cascade to act as if the style attribute were not present, since the style attribute is treated as its own style rule.

Reviewer: ```suggestion
When `revert-rule` is used in a [style attribute](/en-US/docs/Web/HTML/Reference/Global_attributes/style), it causes the cascade to act as if the style attribute were not present. This works because the style attribute is treated as its own style rule.
```

#### HTML

```html
<p style="color: revert-rule">This text uses the stylesheet color.</p>
```

#### CSS

```css hidden
body {
  font-family: system-ui;
}

@supports not (color: revert-rule) {
  body::before {
    content: "Your browser doesn't support the revert-rule keyword yet.";
    background-color: wheat;
    display: block;
    text-align: center;
    padding: 1em;
  }
}
```

```css
p {
  color: green;
}
```

#### Result

{{EmbedLiveSample('Reverting_from_a_style_attribute', '100%', 120)}}

> The paragraph displays in green because `revert-rule` removes the style attribute's declaration from the cascade, and the `p { color: green; }` rule takes effect.

Reviewer: ```suggestion
The paragraph text is green because `revert-rule` causes the cascade to ignore the style attribute's declaration, and the `p` rule takes effect.
```

> ### A chain of revert-rule

Reviewer: ```suggestion
### Chaining multiple `revert-rule` values
```

> The `revert-rule` keyword removes the current rule from the cascade entirely. If the resulting winning rule also uses `revert-rule`, that rule is likewise removed, continuing back through earlier rules.

Reviewer: - might be better to use "ignore" instead of "remove"
- the code and remaining prose will need to be updated - it contains only one `revert-rule`
